// valgen/esm/rules/format-rules/is-email.js
import validatorJS from '@browsery/validator';
import { validator, } from '../../core/index.js';
/**
 * Validates if value is a valid Email
 * @validator isEmail
 */
export function isEmail(options) {
    const emailOptions = {
        allow_display_name: true,
        allow_utf8_local_part: options?.utf8LocalPart,
        ignore_max_length: true,
        allow_ip_domain: options?.allowIpDomain,
    };
    return validator('isEmail', (input, context, _this) => {
        if (typeof input === 'string' &&
            validatorJS.isEmail(input, emailOptions)) {
            if (options?.requireDisplayName) {
                if (!validatorJS.isEmail(input, {
                    ...emailOptions,
                    require_display_name: true,
                })) {
                    context.fail(_this, `Display name for the email is required. Etc. ( Name <me@tempuri.org> ) `, input);
                    return;
                }
            }
            else if (!options?.allowDisplayName &&
                !validatorJS.isEmail(input, {
                    ...emailOptions,
                    allow_display_name: false,
                })) {
                context.fail(_this, `Display name in email is not allowed`, input);
                return;
            }
            if (options?.hostBlacklist &&
                !validatorJS.isEmail(input, {
                    ...emailOptions,
                    host_blacklist: options.hostBlacklist,
                })) {
                context.fail(_this, `Email "{{value}}" is in black-list`, input);
                return;
            }
            if (options?.hostWhitelist &&
                !validatorJS.isEmail(input, {
                    ...emailOptions,
                    host_whitelist: options.hostWhitelist,
                })) {
                context.fail(_this, `Email "{{value}}" is in not white-list`, input);
                return;
            }
            if (options?.blacklistedChars &&
                !validatorJS.isEmail(input, {
                    ...emailOptions,
                    blacklisted_chars: options.blacklistedChars,
                })) {
                context.fail(_this, `Black listed characters (${options.blacklistedChars}) found in name part`, input);
                return;
            }
            return input;
        }
        context.fail(_this, `Value must much required e-mail format`, input);
    }, options);
}
