// valgen/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isUUID4 = exports.isUUID3 = exports.isUUID2 = exports.isUUID1 = exports.isUUID = exports.isURL = exports.isUppercase = exports.isUndefined = exports.isTime = exports.isSWIFT = exports.isString = exports.isPort = exports.isObjectId = exports.isObject = exports.isNumber = exports.isNullish = exports.isNull = exports.isNotNullish = exports.isNotNull = exports.isNotEmpty = exports.isMobilePhone = exports.isMACAddress = exports.isLowercase = exports.isJWT = exports.isISSN = exports.isIPRange = exports.isIP = exports.isInteger = exports.isIBAN = exports.isHexColor = exports.isHex = exports.isFQDN = exports.isETHAddress = exports.isEmpty = exports.isEmail = exports.isEAN = exports.isDefined = exports.isDecimal = exports.isDateString = exports.isDate = exports.isCreditCard = exports.isBtcAddress = exports.isBoolean = exports.isBigint = exports.isBase64 = exports.isAscii = exports.isArray = exports.isAny = exports.isAlphanumeric = exports.isAlpha = void 0;
exports.vg = exports.toTime = exports.toString = exports.toNumber = exports.toInteger = exports.toDateString = exports.toDate = exports.toBoolean = exports.toBigint = exports.toArray = exports.isUUID5 = void 0;
const tslib_1 = require("tslib");
const vg = tslib_1.__importStar(require("./rules/index.js"));
exports.vg = vg;
tslib_1.__exportStar(require("./constants.js"), exports);
tslib_1.__exportStar(require("./core/index.js"), exports);
const isAlpha = vg.isAlpha();
exports.isAlpha = isAlpha;
const isAlphanumeric = vg.isAlphanumeric();
exports.isAlphanumeric = isAlphanumeric;
const isAny = vg.isAny();
exports.isAny = isAny;
const isArray = vg.isArray();
exports.isArray = isArray;
const isAscii = vg.isAscii();
exports.isAscii = isAscii;
const isBase64 = vg.isBase64();
exports.isBase64 = isBase64;
const isBigint = vg.isBigint();
exports.isBigint = isBigint;
const isBoolean = vg.isBoolean();
exports.isBoolean = isBoolean;
const isBtcAddress = vg.isBtcAddress();
exports.isBtcAddress = isBtcAddress;
const isCreditCard = vg.isCreditCard();
exports.isCreditCard = isCreditCard;
const isDate = vg.isDate();
exports.isDate = isDate;
const isDateString = vg.isDateString();
exports.isDateString = isDateString;
const isDecimal = vg.isDecimal();
exports.isDecimal = isDecimal;
const isDefined = vg.isDefined();
exports.isDefined = isDefined;
const isEAN = vg.isEAN();
exports.isEAN = isEAN;
const isEmail = vg.isEmail();
exports.isEmail = isEmail;
const isEmpty = vg.isEmpty();
exports.isEmpty = isEmpty;
const isETHAddress = vg.isETHAddress();
exports.isETHAddress = isETHAddress;
const isFQDN = vg.isFQDN();
exports.isFQDN = isFQDN;
const isHex = vg.isHex();
exports.isHex = isHex;
const isHexColor = vg.isHexColor();
exports.isHexColor = isHexColor;
const isIBAN = vg.isIBAN();
exports.isIBAN = isIBAN;
const isInteger = vg.isInteger();
exports.isInteger = isInteger;
const isIP = vg.isIP();
exports.isIP = isIP;
const isIPRange = vg.isIPRange();
exports.isIPRange = isIPRange;
const isISSN = vg.isISSN();
exports.isISSN = isISSN;
const isJWT = vg.isJWT();
exports.isJWT = isJWT;
const isLowercase = vg.isLowercase();
exports.isLowercase = isLowercase;
const isMACAddress = vg.isMACAddress();
exports.isMACAddress = isMACAddress;
const isMobilePhone = vg.isMobilePhone();
exports.isMobilePhone = isMobilePhone;
const isNotEmpty = vg.isNotEmpty();
exports.isNotEmpty = isNotEmpty;
const isNotNull = vg.isNotNull();
exports.isNotNull = isNotNull;
const isNotNullish = vg.isNotNullish();
exports.isNotNullish = isNotNullish;
const isNull = vg.isNull();
exports.isNull = isNull;
const isNullish = vg.isNullish();
exports.isNullish = isNullish;
const isNumber = vg.isNumber();
exports.isNumber = isNumber;
const isObject = vg.isObject();
exports.isObject = isObject;
const isObjectId = vg.isObjectId();
exports.isObjectId = isObjectId;
const isPort = vg.isPort();
exports.isPort = isPort;
const isString = vg.isString();
exports.isString = isString;
const isSWIFT = vg.isSWIFT();
exports.isSWIFT = isSWIFT;
const isTime = vg.isTime();
exports.isTime = isTime;
const isUndefined = vg.isUndefined();
exports.isUndefined = isUndefined;
const isUppercase = vg.isUppercase();
exports.isUppercase = isUppercase;
const isURL = vg.isURL();
exports.isURL = isURL;
const isUUID = vg.isUUID();
exports.isUUID = isUUID;
const isUUID1 = vg.isUUID(1);
exports.isUUID1 = isUUID1;
const isUUID2 = vg.isUUID(2);
exports.isUUID2 = isUUID2;
const isUUID3 = vg.isUUID(3);
exports.isUUID3 = isUUID3;
const isUUID4 = vg.isUUID(4);
exports.isUUID4 = isUUID4;
const isUUID5 = vg.isUUID(5);
exports.isUUID5 = isUUID5;
const toArray = vg.isArray(isAny, { coerce: true });
exports.toArray = toArray;
const toBigint = vg.isBigint({ coerce: true });
exports.toBigint = toBigint;
const toBoolean = vg.isBoolean({ coerce: true });
exports.toBoolean = toBoolean;
const toDate = vg.isDate({ coerce: true });
exports.toDate = toDate;
const toDateString = vg.isDateString({ coerce: true, precisionMin: 'day' });
exports.toDateString = toDateString;
const toInteger = vg.isInteger({ coerce: true });
exports.toInteger = toInteger;
const toNumber = vg.isNumber({ coerce: true });
exports.toNumber = toNumber;
const toString = vg.isString({ coerce: true });
exports.toString = toString;
const toTime = vg.isTime({ coerce: true });
exports.toTime = toTime;
