// valgen/cjs/rules/type-rules/is-string.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.trimStart = exports.trimEnd = void 0;
exports.isString = isString;
exports.stringReplace = stringReplace;
exports.stringSplit = stringSplit;
exports.trim = trim;
const index_js_1 = require("../../core/index.js");
/**
 * Validates if value is "string".
 * Converts input value to string if coerce option is set to 'true'.
 * @validator isString
 */
function isString(options) {
    return (0, index_js_1.validator)('isString', (input, context, _this) => {
        const coerce = options?.coerce ?? context.coerce;
        let output = input;
        if (output != null && typeof output !== 'string' && coerce) {
            if (typeof output === 'object') {
                if (typeof output.toJSON === 'function')
                    output = output.toJSON();
                else
                    output = JSON.stringify(output);
            }
            else
                output = String(output);
        }
        if (typeof output === 'string')
            return output;
        context.fail(_this, `Value must be a string`, input);
    }, options);
}
function stringReplace(searchValue, replacer) {
    return (0, index_js_1.validator)('stringReplace', (input) => {
        if (input == null)
            return input;
        return String(input).replace(searchValue, replacer);
    });
}
function stringSplit(splitter, limit) {
    return (0, index_js_1.validator)('stringSplit', (input) => {
        if (input == null)
            return input;
        return String(input).split(splitter, limit);
    });
}
/**
 * Removes whitespace from both ends of a string
 * @validator trim
 */
function trim() {
    return (0, index_js_1.validator)('trim', (input) => {
        if (input == null)
            return input;
        return String(input).trim();
    });
}
// *************************************************************
/**
 * Removes whitespace from the end of a string
 * @validator trimEnd
 */
const trimEnd = () => trimEndRule;
exports.trimEnd = trimEnd;
const trimEndRule = (0, index_js_1.validator)('trimEnd', (input) => {
    if (input == null)
        return input;
    return String(input).trimEnd();
});
// *************************************************************
/**
 * Removes whitespace from the beginning of a string
 * @validator trimStart
 */
const trimStart = () => trimStartRule;
exports.trimStart = trimStart;
const trimStartRule = (0, index_js_1.validator)('trimStart', (input) => {
    if (input == null)
        return input;
    return String(input).trimStart();
});
