// valgen/esm/rules/type-rules/is-string.js
import { validator, } from '../../core/index.js';
/**
 * Validates if value is "string".
 * Converts input value to string if coerce option is set to 'true'.
 * @validator isString
 */
export function isString(options) {
    return validator('isString', (input, context, _this) => {
        const coerce = options?.coerce ?? context.coerce;
        let output = input;
        if (output != null && typeof output !== 'string' && coerce) {
            if (typeof output === 'object') {
                if (typeof output.toJSON === 'function')
                    output = output.toJSON();
                else
                    output = JSON.stringify(output);
            }
            else
                output = String(output);
        }
        if (typeof output === 'string')
            return output;
        context.fail(_this, `Value must be a string`, input);
    }, options);
}
export function stringReplace(searchValue, replacer) {
    return validator('stringReplace', (input) => {
        if (input == null)
            return input;
        return String(input).replace(searchValue, replacer);
    });
}
export function stringSplit(splitter, limit) {
    return validator('stringSplit', (input) => {
        if (input == null)
            return input;
        return String(input).split(splitter, limit);
    });
}
/**
 * Removes whitespace from both ends of a string
 * @validator trim
 */
export function trim() {
    return validator('trim', (input) => {
        if (input == null)
            return input;
        return String(input).trim();
    });
}
// *************************************************************
/**
 * Removes whitespace from the end of a string
 * @validator trimEnd
 */
export const trimEnd = () => trimEndRule;
const trimEndRule = validator('trimEnd', (input) => {
    if (input == null)
        return input;
    return String(input).trimEnd();
});
// *************************************************************
/**
 * Removes whitespace from the beginning of a string
 * @validator trimStart
 */
export const trimStart = () => trimStartRule;
const trimStartRule = validator('trimStart', (input) => {
    if (input == null)
        return input;
    return String(input).trimStart();
});
